art-decisions/proposals/2022-11-20-arcaflow-signals.md

```diff
@@ -0,0 +1,211 @@
+# Adding signals to the Arcaflow execution model
+
+## Background
+
+On Friday the 17th of November 2022 we had a very lengthy discussion with the Red Hat Performance & Scale Middleware
+team. They have introduced us to their extremely powerful tool for executing workflows called
+[qDup](https://github.com/Hyperfoil/qDup). This tool not only allows people to write workflows, but also dynamically
+react to outputs of programs and publish signals on an internal messaging bus. Workflow parts can communicate with each
+other even while running, sending and waiting for signals.
+
+Arguably, qDup can do more than Arcaflow at this point. However, it requires a great deal of expertise and forethought
+as it doesn't have the safety guarantees that Arcaflow does, such as schemas or possible static analysis of the code.
+
+It is the aim of this proposal that Arcaflow should receive the signaling mechanism, allowing plugins to exchange
+messages during execution. This proposal incorporates the plan to create data streaming, as signals will allow plugins
+to transmit chunks of data.
+
+## Proposal
+
+Currently, Arcaflow allows the plugin to output only one of many possible outputs. This limits their usefulness,
+as the plugin must finish execution to transmit information.
+
+In this proposal, we transform the execution of Arcaflow by adding the ability to send and receive signals via signal
+channels. Each signal channel will have a schema and is declared by a plugin. Workflow authors can take these signals 
+and pipe them into other plugins that have declared they can receive signals.
```

Comment:
For steps that just do work and report the result, what would be the advantage of using signals for them?
It may just be easier to have the existing input and output option, with signals for events and mid-step data transfers.